### PR TITLE
Adds yolov3 model using tvm to AutowareAuto

### PR DESCRIPTION
**Pull request by daggarwa**

### CODE REVIEW

**Comment by LucaFos** [comment]
Hello @daggarwa, and thanks for the PR. As part of the PR I see 3 models, but only one seems to actually be used. Why is this? We'd rather have the other 2 removed if they are actually not needed.

**Comment by daggarwa** [comment]
> Hello @daggarwa, and thanks for the PR. As part of the PR I see 3 models, but only one seems to be used. Why is this? We'd rather have the other 2 removed if they are actually not needed.

Hi @LucaFos Please wait. The PR is not finished yet. I will clean up stuff and submit for review soon.

**Comment by daggarwa** [comment]
@LucaFos I think this PR is ready for a first round of review. You can find instructions here : perception/camera_obstacle_detection/yolo_v3/tensorflow_fp32_coco/README.md on how to build the model and run the example pipeline. Please let me know if you face any issues or have any questions. Thanks.

**Comment by ambroise-arm** [comment]
Current status:

@daggarwa to rebase on tip of master in order to get the update of the TVM version (and keep the master version of install_tvm.sh), and fix the `shell-lint` errors in the CI.

**Comment by daggarwa** [comment]
> Current status:
> 
> @daggarwa to rebase on tip of master in order to get the update of the TVM version (and keep the master version of install_tvm.sh), and fix the `shell-lint` errors in the CI.

@ambroise-arm  I will be able to have a look at this in this week and resolve this.

**Comment by daggarwa** [comment]
> > Current status:
> > @daggarwa to rebase on tip of master in order to get the update of the TVM version (and keep the master version of install_tvm.sh), and fix the `shell-lint` errors in the CI.
> 
> @ambroise-arm I will be able to have a look at this in this week and resolve this.

I am working on this today. Will try to resolve it today

**Comment by daggarwa** [comment]
@ambroise-arm I am getting this error with the latest version of TVM :  
```
An error occurred during the execution of TVM.
For more information, please see: https://tvm.apache.org/docs/errors.html
---------------------------------------------------------------
  Check failed: (lib_handle_ != nullptr) is false: Failed to load dynamic shared library ./deploy_lib.so ./deploy_lib.so: cannot open shared object file: No such file or directory
Stack trace:
  0: tvm::runtime::DSOLibrary::Load(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
  1: tvm::runtime::CreateDSOLibraryObject(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >)
  2: std::_Function_handler<void (tvm::runtime::TVMArgs, tvm::runtime::TVMRetValue*), tvm::runtime::{lambda(tvm::runtime::TVMArgs, tvm::runtime::TVMRetValue*)#1}>::_M_invoke(std::_Any_data const&, tvm::runtime::TVMArgs&&, tvm::runtime::TVMRetValue*&&)
  3: tvm::runtime::Module::LoadFromFile(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
  4: main
        at /home/divya.aggarwal/ade-home/modelzoo/perception/camera_obstacle_detection/yolo_v3/tensorflow_fp32_coco/example_pipeline/main.cpp:174
  5: __libc_start_main
  6: _start
  7: 0xffffffffffffffff
```

Would you happen to know anything about this?

**Comment by LucaFos** [comment]
> 

Hi @daggarwa, which command gives you that error?

**Comment by LucaFos** [comment]
> @ambroise-arm I am getting this error with the latest version of TVM after rebasing :
> 
> ```
> An error occurred during the execution of TVM.
> For more information, please see: https://tvm.apache.org/docs/errors.html
> ---------------------------------------------------------------
>   Check failed: (lib_handle_ != nullptr) is false: Failed to load dynamic shared library ./deploy_lib.so ./deploy_lib.so: cannot open shared object file: No such file or directory
> Stack trace:
>   0: tvm::runtime::DSOLibrary::Load(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
>   1: tvm::runtime::CreateDSOLibraryObject(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >)
>   2: std::_Function_handler<void (tvm::runtime::TVMArgs, tvm::runtime::TVMRetValue*), tvm::runtime::{lambda(tvm::runtime::TVMArgs, tvm::runtime::TVMRetValue*)#1}>::_M_invoke(std::_Any_data const&, tvm::runtime::TVMArgs&&, tvm::runtime::TVMRetValue*&&)
>   3: tvm::runtime::Module::LoadFromFile(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
>   4: main
>         at /home/divya.aggarwal/ade-home/modelzoo/perception/camera_obstacle_detection/yolo_v3/tensorflow_fp32_coco/example_pipeline/main.cpp:174
>   5: __libc_start_main
>   6: _start
>   7: 0xffffffffffffffff
> ```
> 
> Would you happen to know anything about this?

Hi @daggarwa, which command gives you the error?

**Comment by daggarwa** [comment]
> > @ambroise-arm I am getting this error with the latest version of TVM after rebasing :
> > ```
> > An error occurred during the execution of TVM.
> > For more information, please see: https://tvm.apache.org/docs/errors.html
> > ---------------------------------------------------------------
> >   Check failed: (lib_handle_ != nullptr) is false: Failed to load dynamic shared library ./deploy_lib.so ./deploy_lib.so: cannot open shared object file: No such file or directory
> > Stack trace:
> >   0: tvm::runtime::DSOLibrary::Load(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
> >   1: tvm::runtime::CreateDSOLibraryObject(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >)
> >   2: std::_Function_handler<void (tvm::runtime::TVMArgs, tvm::runtime::TVMRetValue*), tvm::runtime::{lambda(tvm::runtime::TVMArgs, tvm::runtime::TVMRetValue*)#1}>::_M_invoke(std::_Any_data const&, tvm::runtime::TVMArgs&&, tvm::runtime::TVMRetValue*&&)
> >   3: tvm::runtime::Module::LoadFromFile(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
> >   4: main
> >         at /home/divya.aggarwal/ade-home/modelzoo/perception/camera_obstacle_detection/yolo_v3/tensorflow_fp32_coco/example_pipeline/main.cpp:174
> >   5: __libc_start_main
> >   6: _start
> >   7: 0xffffffffffffffff
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Would you happen to know anything about this?
> 
> Hi @daggarwa, which command gives you the error? Edit: I guess https://github.com/autowarefoundation/modelzoo/pull/57/files#diff-2d5ed8b1bab389029af8f4026b52acf0df48764dfc5b94bab9d92a91892f2a8cR18

@LucaFos Yes exactly. This is giving the error: 
```
docker run -it --rm -v /tmp/.X11-unix:/tmp/.X11-unix:rw -v /home/divya.aggarwal/.Xauthority:/home/divya.aggarwal/.Xauthority:rw -e XAUTHORITY=/home/divya.aggarwal/.Xauthority -e DISPLAY=:1 --net=host -v /home/divya.aggarwal/ade-home/modelzoo/perception/camera_obstacle_detection/yolo_v3/tensorflow_fp32_coco/:/home/divya.aggarwal/ade-home/modelzoo/perception/camera_obstacle_detection/yolo_v3/tensorflow_fp32_coco/ -w /home/divya.aggarwal/ade-home/modelzoo/perception/camera_obstacle_detection/yolo_v3/tensorflow_fp32_coco//example_pipeline/build --entrypoint '' autoware/model-zoo-tvm-cli:local bash -c 'LD_LIBRARY_PATH=:/usr/local/lib/ ./example_pipeline'
```
Any idea how do I remedy that?

**Comment by LucaFos** [comment]
> > > @ambroise-arm I am getting this error with the latest version of TVM after rebasing :
> > > ```
> > > An error occurred during the execution of TVM.
> > > For more information, please see: https://tvm.apache.org/docs/errors.html
> > > ---------------------------------------------------------------
> > >   Check failed: (lib_handle_ != nullptr) is false: Failed to load dynamic shared library ./deploy_lib.so ./deploy_lib.so: cannot open shared object file: No such file or directory
> > > Stack trace:
> > >   0: tvm::runtime::DSOLibrary::Load(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
> > >   1: tvm::runtime::CreateDSOLibraryObject(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >)
> > >   2: std::_Function_handler<void (tvm::runtime::TVMArgs, tvm::runtime::TVMRetValue*), tvm::runtime::{lambda(tvm::runtime::TVMArgs, tvm::runtime::TVMRetValue*)#1}>::_M_invoke(std::_Any_data const&, tvm::runtime::TVMArgs&&, tvm::runtime::TVMRetValue*&&)
> > >   3: tvm::runtime::Module::LoadFromFile(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
> > >   4: main
> > >         at /home/divya.aggarwal/ade-home/modelzoo/perception/camera_obstacle_detection/yolo_v3/tensorflow_fp32_coco/example_pipeline/main.cpp:174
> > >   5: __libc_start_main
> > >   6: _start
> > >   7: 0xffffffffffffffff
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > Would you happen to know anything about this?
> > 
> > 
> > Hi @daggarwa, which command gives you the error? Edit: I guess https://github.com/autowarefoundation/modelzoo/pull/57/files#diff-2d5ed8b1bab389029af8f4026b52acf0df48764dfc5b94bab9d92a91892f2a8cR18
> 
> @LucaFos Yes exactly. This is giving the error:
> 
> ```
> docker run -it --rm -v /tmp/.X11-unix:/tmp/.X11-unix:rw -v /home/divya.aggarwal/.Xauthority:/home/divya.aggarwal/.Xauthority:rw -e XAUTHORITY=/home/divya.aggarwal/.Xauthority -e DISPLAY=:1 --net=host -v /home/divya.aggarwal/ade-home/modelzoo/perception/camera_obstacle_detection/yolo_v3/tensorflow_fp32_coco/:/home/divya.aggarwal/ade-home/modelzoo/perception/camera_obstacle_detection/yolo_v3/tensorflow_fp32_coco/ -w /home/divya.aggarwal/ade-home/modelzoo/perception/camera_obstacle_detection/yolo_v3/tensorflow_fp32_coco//example_pipeline/build --entrypoint '' autoware/model-zoo-tvm-cli:local bash -c 'LD_LIBRARY_PATH=:/usr/local/lib/ ./example_pipeline'
> ```
> 
> Any idea how do I remedy that?

Hi again @daggarwa, I just tried your code. First I deleted the contents of

> ${MODEL_DIR}/example_pipeline/build

then I followed the steps in

> perception/camera_obstacle_detection/yolo_v3/tensorflow_fp32_coco/example_pipeline/scripts/yolov3-416.sh

except for the last command, where instead of

`docker run      -it --rm      -v /tmp/.X11-unix:/tmp/.X11-unix:rw      -v ${HOME}/.Xauthority:${HOME}/.Xauthority:rw      -e XAUTHORITY=${HOME}/.Xauthority      -e DISPLAY=$DISPLAY      --net=host      -v ${MODEL_DIR}:${MODEL_DIR}      -w ${MODEL_DIR}/example_pipeline/build      --entrypoint ""      autoware/model-zoo-tvm-cli:local          bash -c "LD_LIBRARY_PATH=$LD_LIBRARY_PATH:/usr/local/lib/ ./example_pipeline"
`

I simply ran

`docker run      -it --rm      -v /tmp/.X11-unix:/tmp/.X11-unix:rw      -v ${HOME}/.Xauthority:${HOME}/.Xauthority:rw      -e XAUTHORITY=${HOME}/.Xauthority      -e DISPLAY=$DISPLAY      --net=host      -v ${MODEL_DIR}:${MODEL_DIR}      -w ${MODEL_DIR}/example_pipeline/build      --entrypoint ""      autoware/model-zoo-tvm-cli:local       ./example_pipeline`

and it seems to work for me.

**Comment by daggarwa** [comment]
> > Any idea how do I remedy that?
> 
> Hi again @daggarwa, I just tried your code. First I deleted the contents of
> 
> `${MODEL_DIR}/example_pipeline/build`
> 
> then I followed the steps in
> 
> `perception/camera_obstacle_detection/yolo_v3/tensorflow_fp32_coco/example_pipeline/scripts/yolov3-416.sh`
> 
> except for the last command, where instead of
> 
> `$ docker run -it --rm -v /tmp/.X11-unix:/tmp/.X11-unix:rw -v ${HOME}/.Xauthority:${HOME}/.Xauthority:rw -e XAUTHORITY=${HOME}/.Xauthority -e DISPLAY=$DISPLAY --net=host -v ${MODEL_DIR}:${MODEL_DIR} -w ${MODEL_DIR}/example_pipeline/build --entrypoint "" autoware/model-zoo-tvm-cli:local bash -c "LD_LIBRARY_PATH=$LD_LIBRARY_PATH:/usr/local/lib/ ./example_pipeline"`
> 
> I simply ran
> 
> `$ docker run -it --rm -v /tmp/.X11-unix:/tmp/.X11-unix:rw -v ${HOME}/.Xauthority:${HOME}/.Xauthority:rw -e XAUTHORITY=${HOME}/.Xauthority -e DISPLAY=$DISPLAY --net=host -v ${MODEL_DIR}:${MODEL_DIR} -w ${MODEL_DIR}/example_pipeline/build --entrypoint "" autoware/model-zoo-tvm-cli:local ./example_pipeline`
> 
> and it seems to work for me.

We had an offline chat to resolve this. This happened due to the model file not being present on my system after a fresh clone of the repo. I have pushed all the changes discussed now.

File: perception/camera_obstacle_detection/yolo_v3/tensorflow_fp32_coco/definition.yaml [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+        - 26
+        - 255
+
+    - name: conv2d_74
+      description:
+      shape:
+        - 1
+        - 52
+        - 52
+        - 255
```
Comment (LucaFos) [inline perception/camera_obstacle_detection/yolo_v3/tensorflow_fp32_coco/definition.yaml:43]: missing newline at the end of file

File: perception/camera_obstacle_detection/yolo_v3/tensorflow_fp32_coco/execute_model/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+# Copyright (c) 2020, Arm Limited and Contributors. All rights reserved.
+#
+# SPDX-License-Identifier: Apache-2.0
+#
+
+cmake_minimum_required(VERSION 3.0)
+set(CMAKE_CXX_STANDARD_REQUIRED ON)
+set(CMAKE_CXX_STANDARD 14)
+
+project( example_pipeline )
```
Comment (LucaFos) [inline perception/camera_obstacle_detection/yolo_v3/tensorflow_fp32_coco/execute_model/CMakeLists.txt:11]: Can you change the name of the project here?

File: perception/camera_obstacle_detection/yolo_v3/tensorflow_fp32_coco/execute_model/main.cpp [hunk 3]
```diff
@@ -0,0 +1,434 @@
...
+#define NETWORK_OUTPUT_WIDTH_2 26
+
+// network output dimensions
+#define NETWORK_OUTPUT_HEIGHT_3 52
+#define NETWORK_OUTPUT_WIDTH_3 52
+
+// minimum confidence score by which to filter the output detections
+#define SCORE_THRESHOLD 0.5
+
+#define NMS_THRESHOLD 0.45
```
Comment (LucaFos) [inline perception/camera_obstacle_detection/yolo_v3/tensorflow_fp32_coco/execute_model/main.cpp:64]: can you add a comment about NSM_THRESHOLD?

File: perception/camera_obstacle_detection/yolo_v3/tensorflow_fp32_coco/execute_model/main.cpp [hunk 4]
```diff
@@ -0,0 +1,434 @@
...
+
+/// Map for storing the class and detections in the image
+std::map<int, std::vector<BoundingBox>> bbox_map{};
+static const int NETWORK_OUTPUT_WIDTH[3] = {NETWORK_OUTPUT_WIDTH_1, NETWORK_OUTPUT_WIDTH_2, NETWORK_OUTPUT_WIDTH_3};
+static const int NETWORK_OUTPUT_HEIGHT[3] = {NETWORK_OUTPUT_HEIGHT_1, NETWORK_OUTPUT_HEIGHT_2, NETWORK_OUTPUT_HEIGHT_3};
+
+
+/// \brief gets the amount of overlap between the two bounding boxes
+/// \return returns the overlap amount
+
```
Comment (LucaFos) [inline perception/camera_obstacle_detection/yolo_v3/tensorflow_fp32_coco/execute_model/main.cpp:93]: can you remove this empty line? Just for consistency.

File: perception/camera_obstacle_detection/yolo_v3/tensorflow_fp32_coco/execute_model/main.cpp [hunk 5]
```diff
@@ -0,0 +1,434 @@
...
+// filename of the image on which to run the inference
+#define IMAGE_FILENAME "test_image_0.jpg"
+
+// Name of the window to display the detection results
+#define DISPLAY_WINDOW_NAME "YOLO Output"
+
+/// Struct for storing detections in image with confidence scores
+struct BoundingBox
+{
+
```
Comment (LucaFos) [inline perception/camera_obstacle_detection/yolo_v3/tensorflow_fp32_coco/execute_model/main.cpp:75]: why do we have these extra empty lines? Same question also for lines 151 and 155

**Review by LucaFos — CHANGES_REQUESTED** [review]
See the inline comments, plus a question: is there a reason why the .onnx file is called yolov3-416.onnx instead of yolov3.onnx?

File: perception/camera_obstacle_detection/yolo_v3/tensorflow_fp32_coco/execute_model/CMakeLists.txt (hunk 2 above)
Comment (daggarwa) [reply perception/camera_obstacle_detection/yolo_v3/tensorflow_fp32_coco/execute_model/CMakeLists.txt]: Changed

File: perception/camera_obstacle_detection/yolo_v3/tensorflow_fp32_coco/execute_model/main.cpp (hunk 3 above)
Comment (daggarwa) [reply perception/camera_obstacle_detection/yolo_v3/tensorflow_fp32_coco/execute_model/main.cpp:76]: Added

File: perception/camera_obstacle_detection/yolo_v3/tensorflow_fp32_coco/execute_model/main.cpp (hunk 5 above)
Comment (daggarwa) [reply perception/camera_obstacle_detection/yolo_v3/tensorflow_fp32_coco/execute_model/main.cpp]: Done removed the extra empty lines

**Comment by daggarwa** [comment]
> See the inline comments, plus a question: is there a reason why the .onnx file is called yolov3-416.onnx instead of yolov3.onnx?

@LucaFos There are different models for yolov3 based on input size of image: 
```
YOLOv3-320
YOLOv3-416
YOLOv3-608
YOLOv3-tiny
YOLOv3-spp
```
Reference here: https://pjreddie.com/darknet/yolo/ 

I ported the one which takes in an image of size 416 * 416 in this PR hence I specified the 416 in the onnx model file name. Hope this helps.

**Review by LucaFos — APPROVED** [review]

**Pull request merged**